### ISSUE TITLE
what does the value of anchors mean?

### ISSUE THREAD

**Issue by mengce97** [microsoft/singleshotpose#80]
for example 
anchors = 1.4820, 2.2412, 2.0501, 3.1265, 2.3946, 4.6891, 3.1018, 3.9910, 3.4879, 5.8851

what does these vaule mean?
how can i get these value?

thank u for replying

**Comment by btekin** [comment]
1. Please refer to this part of the paper for an explanation on the anchor values:

> When  multiple  objects  are  located  close  to  each  other in  the  3D  scene,  they  are  more  likely  to  appear  close  together in the images or be occluded by each other. In these cases,  certain cells might contain multiple objects.  To be able to predict the pose of such multiple objects that lie in the same cell, we allow up to5candidates per cell and there-fore predict five sets of control points per cell.  This essentially means that we assumed that at most 5 objects could occlude each other in a single grid cell. This is a reasonable assumption to make in practical pose estimation scenarios.As in [29], we precompute with k-means, five anchor boxes that define the size, ie. the width and height of a 2D rectangle tightly fitted to a masked region around the object in the image.   During training,  we assign whichever anchor box has the most similar size to the current object as the responsible one to predict the 2D coordinates for that object.

Those values correspond to, therefore, "the width and height of a 2D rectangle tightly fitted to a masked region around the object". These values are pre-computed with k-means as explained above and normalized by the grid size (final feature map). You could maybe refer to the following code to generate anchors for your own dataset: https://github.com/AlexeyAB/darknet/blob/master/scripts/gen_anchors.py

2. Sqrt is just used to have a more balanced loss function for confidence, coordinate and classification losses. I would assume, the convergence would still be reasonable without it, but haven't tested this.

3. Initially we set the confidence mask to noobject_scale for all cells. 

`conf_mask   = torch.ones(nB, nA, nH, nW) * noobject_scale`

Then we set it to zero at the areas the confidence computed on the fly is above a threshold. 

`conf_mask[b][cur_confs>sil_thresh] = 0`

And finally, w e set the confidence mask to object_scale at the cell where we know the object is present. 

`conf_mask[b][best_n][gj0][gi0]  = object_scale`

This is to give more emphasis and have more precision in where the object is actually located.  By setting cells around the object to 0, we prevent predicting high confidence values in those regions where a confusion is more likely (because of closeness to the object) and eventually this gives more precision.
